minigalaxy/installer.py

```diff
@@ -106,51 +107,82 @@ def make_tmp_dir(game):
-
-        # It's possible to set install dir as argument before installation
-        command = ["env", "WINEPREFIX={}".format(prefix_dir), "wine", installer,
-                   "/dir={}".format(temp_dir), "/VERYSILENT"]
-    process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    process.wait()
-    stdout, stderr = process.communicate()
-    stdout = stdout.decode("utf-8")
-    stderr = stderr.decode("utf-8")
-    if (process.returncode not in [0, 1]) or \
-       (process.returncode in [1] and "(attempting to process anyway)" not in stderr):
-        error_message = _("The installation of {} failed. Please try again.").format(installer)
+        err_msg = extract_windows(game, installer, temp_dir)
+    return err_msg
+
+
+def extract_linux(installer, temp_dir):
+    err_msg = ""
+    command = ["unzip", "-qq", installer, "-d", temp_dir]
+    stdout, stderr, exitcode = _exe_cmd(command)
+    if (exitcode not in [0]) and \
+       (exitcode not in [1] and "(attempting to process anyway)" not in stderr):
+        err_msg = _("The installation of {} failed. Please try again.").format(installer)
     elif len(os.listdir(temp_dir)) == 0:
-        error_message = _("{} could not be unzipped.".format(installer))
-    return error_message
+        err_msg = _("{} could not be unzipped.".format(installer))
+    return err_msg
+
 
+def extract_windows(game, installer, temp_dir):
+    err_msg = extract_by_innoextract(installer, temp_dir)
+    if err_msg:
+        err_msg = extract_by_wine(game, installer, temp_dir)
+    return err_msg
+
+
+def extract_by_innoextract(installer, temp_dir):
+    err_msg = ""
+    if shutil.which("innoextract"):
+        cmd = ["innoextract", installer, "-d", temp_dir]
+        stdout, stderr, exitcode = _exe_cmd(cmd)
+        if exitcode not in [0]:
+            err_msg = _("Innoextract extraction failed.")
+        else:
+            inno_app_dir = os.path.join(temp_dir, "app")
+            if os.path.isdir(inno_app_dir):
+                _mv(inno_app_dir, temp_dir)
+            innoextract_unneeded_dirs = ["__redist", "tmp", "commonappdata", "app"]
+            for unneeded_dir in innoextract_unneeded_dirs:
+                unneeded_dir_full_path = os.path.join(temp_dir, unneeded_dir)
+                if os.path.isdir(unneeded_dir_full_path):
+                    shutil.rmtree(unneeded_dir_full_path)
+    else:
+        err_msg = _("Innoextract not installed.")
+    return err_msg
 
-def move_and_overwrite(game, temp_dir, target_dir):
+
+def extract_by_wine(game, installer, temp_dir):
+    err_msg = ""
+    # Set the prefix for Windows games
+    prefix_dir = os.path.join(game.install_dir, "prefix")
+    if not os.path.exists(prefix_dir):
+        os.makedirs(prefix_dir, mode=0o755)
+    # It's possible to set install dir as argument before installation
+    command = ["env", "WINEPREFIX={}".format(prefix_dir), "wine", installer, "/dir={}".format(temp_dir), "/VERYSILENT"]
```

Comment:
This shouldn't use a temp dir, because it breaks the shortcuts which are made. I would however, not require this to be done within this PR.